### PR TITLE
ref(saml): Refactor to use provider helper flow

### PR DESCRIPTION
**Pull request by evanpurkhiser**
This is a large refactoring of the SAML2 authentication provider to use the AuthHelper, which reduces a *lot* of boiler plate code and helps to provide a really solid separation of concerns.

 * The AuthHelper handles the authentication and setup of authentication as a 'piepline'. It does this by keeping state in the session.  Previously the authentication callback (the Assertion Consumer Service) completely ran out-of-band to the helper. The result of this was that the helper would never finish the pipeline. This refactor puts the ACS endpoint into the pipeline, as a consequence of this things become simpler:

   - The ACS endpoint no longer needs to do *any* type of authentication for the user, other than verifying the SAML assertion.

   - The ACS endpoint simply binds the SAML attributes from the response into the helpers state, build_identity takes care of the rest of the work for providing needed details to authenticate the user.

     Finally the helper handles the *entire* authentication flow which includes:

     + Linking the users account to the identity
     + Creating a new account for the user should they not have one (this
       also correctly uses the first and last name attributes now).

 * As part of this refactoring, We've made a clean abstraction between the SAML provider plugins (okta, onelogin, auth0) and the base saml2 provider.

   - The provider is now *documented* with the expected state during the setup pipeline.

   - Configuration forms have been moved out of the saml2 authentication module and into the plugins module themselves.

   - Provider static methods have been removed in favor of just making them module level functions.

   - Various 'extract' static methods are now simply baked into the caller, as they added very little value in terms of abstract by separation.

I'll have a follow up PR for the saml2 plugin that implements some of the forms here. I'd like to fix those up over there, since some of them were implemented in places they don't need to be (onelogin had the totally custom forms).

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Changes require @getsentry/security sign-off</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

### Security concerns found

- src/sentry/auth/helper.py
- src/sentry/auth/providers/saml2.py


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/auth/providers/saml2.py
Comment (mattrobenolt) [inline src/sentry/auth/providers/saml2.py]: I don't think you can actually do this `.format()` here since `ERR_SAML_FAILED` isn't a string. It's a translation object. Usually these are used with just % formatting, but I'd just double check this.

**Comment by EvanPurkhiser** [comment]
Note that this does include commits from https://github.com/getsentry/sentry/pull/6173, which should be merged first.

File: src/sentry/auth/providers/saml2.py
Comment (EvanPurkhiser) [reply src/sentry/auth/providers/saml2.py]: This does work, actually the `%` formatting wasn't working right: https://github.com/getsentry/sentry/pull/6173#discussion_r140884273

File: src/sentry/auth/providers/saml2.py
Comment (mattrobenolt) [inline src/sentry/auth/providers/saml2.py]: Check `request.user.is_authenticated()` instead.

Comment (EvanPurkhiser) [reply src/sentry/auth/providers/saml2.py]: ccd964a

File: src/sentry/auth/providers/saml2.py
Comment (mattrobenolt) [inline src/sentry/auth/providers/saml2.py]: If you use `filter(None, name)` this would be more correct imo, since this will just filter out all falsey values rather than strictly Nones. Like, if last name were an empty string value.

```python
>>> filter(None, [None, 1, 0, ''])
[1]
```

**Review by mattrobenolt — APPROVED** [review]
From a high level, I like the cleanup and I think this is a bit more clear. I can't speak for the semantics going on, but assuming you have tested all of that thoroughly.

**Comment by EvanPurkhiser** [comment]
Thanks @mattrobenolt. I'll be waiting to get all the changes in the plugin merged in first (PRs coming soon) before merging this in, so we can get the roll to getsentry synced up with bumping the plugin.

File: src/sentry/auth/providers/saml2.py
Comment (EvanPurkhiser) [reply src/sentry/auth/providers/saml2.py]: perfect

**Comment by EvanPurkhiser** [comment]
I added one more commit [30f7247] here as a follow up to https://github.com/getsentry/sentry/pull/6213

The generic SAML plugin allows the users to configure the attribute mapping, in which case, it might be incorrect and the identity won't be created correctly.

**Pull request merged**
